guidance/rust_code/src/graph/mod.rs

```diff
@@ -1,402 +1,282 @@
-        .split(",") // Remove commas and store remaining information in Vec
-        .map(|s| s.parse().unwrap()) // Convert vec of strings to integers
-        .collect::<Vec<f64>>();
+#[derive(Debug)]
+pub struct Graph {
+    pub vertices: Vec<Vertex>,
+    pub edges: Vec<Edge>,
+    pub connection_matrix: DMatrix<Option<EdgeIndex>> // D stands for dynamic
+}
 
-    let latitude = gps_data[0];
-    let longitude = gps_data[1];
+impl Vertices for &Graph {
+    fn vertices<'a, 'b>(&'a self, graph: &'b Graph) -> Vec<&'b Vertex> {
+        graph.vertices.iter().collect()
+    }
+}
 
-    (latitude, longitude)
+impl Edges for &Graph {
+    fn edges<'a, 'b>(&'a self, graph: &'b Graph) -> Vec<&'b Edge> {
+        graph.edges.iter().collect()
+    }
 }
 
 pub(self) fn connect_vertices_with_edges(
     edges: Vec<Edge>, 
     vertices: Vec<Vertex>
 ) -> Graph {
-    let connection_matrix = vec![vec![None; vertices.len()]; vertices.len()];
-    let mut graph = Graph {
-        edges,
-        vertices,
-        connection_matrix
-    };
+
+    let vertices_len = vertices.len();
+
+    let mut connection_matrix = DMatrix::from_vec(
+        vertices_len, // number of rows
+        vertices_len, // number of columns
+        vec![None; vertices_len * vertices_len] // initializing the array to None
+    );
     
-    for edge_number in 0..graph.edges.len() {
-        let start_of_edge = graph.edges[edge_number].gps_points.first();
-        let end_of_edge = graph.edges[edge_number].gps_points.last();
+    for (edege_index, edge) in edges.iter().enumerate() {
+        let start_of_edge = edge.points.first().unwrap();
+        let end_of_edge = edge.points.last().unwrap();
         let mut start_vertex_index = None;
         let mut end_vertex_index = None;
 
-        for vertex_number in 0..graph.vertices.len() {
-            let vertex = &graph.vertices[vertex_number];
-            if Some(&vertex.gps_point) == start_of_edge {
-                start_vertex_index = Some(vertex_number as usize);
+        for (vertex_index, vertex) in vertices.iter().enumerate() {
+            if vertex.point.gps == start_of_edge.gps {
+                start_vertex_index = Some(vertex_index);
             }
-            else if Some(&vertex.gps_point) ==  end_of_edge {
-                end_vertex_index = Some(vertex_number as usize);
+            else if vertex.point.gps ==  end_of_edge.gps {
+                end_vertex_index = Some(vertex_index);
             }
         }
 
         let start_vertex_index = start_vertex_index.unwrap();
         let end_vertex_index = end_vertex_index.unwrap();
 
-        graph.connection_matrix[start_vertex_index][end_vertex_index] = Some(edge_number);
-        graph.connection_matrix[end_vertex_index][start_vertex_index] = Some(edge_number);
+        connection_matrix[(start_vertex_index, end_vertex_index)] = Some(EdgeIndex(edege_index));
+        connection_matrix[(end_vertex_index, start_vertex_index)] = Some(EdgeIndex(edege_index));
```

Comment:
Again, would like to see more comments on a large block of code like this

Reply:
Fixed.